[RPC Framework] Add _script_module_reducer unconditionally for RecursiveScriptModule in RPC pickler

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58020 [RPC Framework] Add _script_module_reducer unconditionally for RecursiveScriptModule in RPC pickler**

Previously there is no RPC pickler for `RecursiveScriptModule`. Although it is a subclass of `ScriptModule`, the reducer of `ScriptModule` is not triggered for `RecursiveScriptModule` when a script remote module is sent over RPC.

This PR checkpoints the investigation of #58274, which makes sure that a RPC pickler is invoked here. This still cannot fix `test_send_remote_module_over_the_wire_script`. Will revisit this bug once there is a feature request from users.

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

Commit:
[RPC Framework] Add _script_module_reducer unconditionally in RPC pickler

This still cannot fix `test_send_remote_module_over_the_wire_script`.

If this test is re-enabled:

```
======================================================================
ERROR: test_send_remote_module_over_the_wire_script (test_process_group_agent.ProcessGroupThreeWorkersRemoteModuleTestWithSpawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 523, in test_send_remote_module_over_the_wire_script
    attrs = rpc.rpc_sync(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 733, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 643, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/internal.py", line 171, in serialize
    p.dump(obj)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/jit/_script.py", line 663, in __getstate__
    raise pickle.PickleError(
_pickle.PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.
```

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Add _script_module_reducer unconditionally in RPC pickler"


This still cannot fix `test_send_remote_module_over_the_wire_script`.

If this test is re-enabled:

```
======================================================================
ERROR: test_send_remote_module_over_the_wire_script (test_process_group_agent.ProcessGroupThreeWorkersRemoteModuleTestWithSpawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 523, in test_send_remote_module_over_the_wire_script
    attrs = rpc.rpc_sync(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 733, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 643, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/internal.py", line 171, in serialize
    p.dump(obj)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/jit/_script.py", line 663, in __getstate__
    raise pickle.PickleError(
_pickle.PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.
```

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Add both _script_module_reducer and _script_module_reducer unconditionally in RPC pickler"


This still cannot fix `test_send_remote_module_over_the_wire_script`.

If this test is re-enabled:

```
======================================================================
ERROR: test_send_remote_module_over_the_wire_script (test_process_group_agent.ProcessGroupThreeWorkersRemoteModuleTestWithSpawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 523, in test_send_remote_module_over_the_wire_script
    attrs = rpc.rpc_sync(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 733, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 643, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/internal.py", line 171, in serialize
    p.dump(obj)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/jit/_script.py", line 663, in __getstate__
    raise pickle.PickleError(
_pickle.PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.
```

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

Commit:
Update on "[WIP] [RPC Framework] Add both _script_module_reducer and _script_module_reducer unconditionally in RPC pickler"


This still cannot fix `test_send_remote_module_over_the_wire_script`.

If this test is re-enabled:

```
======================================================================
ERROR: test_send_remote_module_over_the_wire_script (test_process_group_agent.ProcessGroupThreeWorkersRemoteModuleTestWithSpawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 523, in test_send_remote_module_over_the_wire_script
    attrs = rpc.rpc_sync(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 733, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 643, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/internal.py", line 171, in serialize
    p.dump(obj)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/jit/_script.py", line 663, in __getstate__
    raise pickle.PickleError(
_pickle.PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.
```

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

SciPioneer commented:
It seems that this change fails a TensorPipe test `test_create_script_module_on_remote`.

Example failure: https://app.circleci.com/pipelines/github/pytorch/pytorch/321756/workflows/9f9a0a52-8c19-46b7-b2f0-0bf8f2caaf12/jobs/13443779
```
Traceback (most recent call last):
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/testing/_internal/distributed/rpc/jit/rpc_test.py", line 961, in test_create_script_module_on_remote
    dst_name, MyScriptModule, args=(self.rank,)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/distributed/rpc/api.py", line 734, in rpc_sync
    return fut.wait()
RuntimeError: PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.

At:
  /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/jit/_script.py(666): __getstate__
  /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/distributed/rpc/internal.py(133): serialize
  /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/distributed/rpc/internal.py(186): serialize
 ```

Commit:
Update on "[RPC Framework] Add _script_module_reducer unconditionally for both ScriptModule and RecursiveScriptModule in RPC pickler"


Previously there is no RPC pickler for `RecursiveScriptModule`. Although it is a subclass of `ScriptModule`, the reducer of `ScriptModule` is not triggered for `RecursiveScriptModule`.

TODO: This still cannot fix `test_send_remote_module_over_the_wire_script`, because the pickling of an RRef is involved, but its RPC pickler is not  called.


If the test `test_send_remote_module_over_the_wire_script` is re-enabled:

If this test is re-enabled:

```
======================================================================
ERROR: test_send_remote_module_over_the_wire_script (test_process_group_agent.ProcessGroupThreeWorkersRemoteModuleTestWithSpawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 523, in test_send_remote_module_over_the_wire_script
    attrs = rpc.rpc_sync(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 733, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 643, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/internal.py", line 171, in serialize
    p.dump(obj)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/jit/_script.py", line 663, in __getstate__
    raise pickle.PickleError(
_pickle.PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.
```

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Add _script_module_reducer unconditionally for both ScriptModule and RecursiveScriptModule in RPC pickler"


Previously there is no RPC pickler for `RecursiveScriptModule`. Although it is a subclass of `ScriptModule`, the reducer of `ScriptModule` is not triggered for `RecursiveScriptModule`.

TODO: This still cannot fix `test_send_remote_module_over_the_wire_script`, because the pickling of an RRef is involved, but its RPC pickler is not  called.


If the test `test_send_remote_module_over_the_wire_script` is re-enabled:

If this test is re-enabled:

```
======================================================================
ERROR: test_send_remote_module_over_the_wire_script (test_process_group_agent.ProcessGroupThreeWorkersRemoteModuleTestWithSpawn)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 391, in wrapper
    self._join_processes(fn)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 583, in _join_processes
    self._check_return_codes(elapsed_time)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 626, in _check_return_codes
    raise RuntimeError(error)
RuntimeError: Process 0 exited with error code 10 and exception:
Traceback (most recent call last):
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 509, in run_test
    getattr(self, test_name)()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/common_distributed.py", line 393, in wrapper
    fn()
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/dist_utils.py", line 100, in new_test_method
    return_value = old_test_method(self, *arg, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/testing/_internal/distributed/nn/api/remote_module_test.py", line 523, in test_send_remote_module_over_the_wire_script
    attrs = rpc.rpc_sync(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 79, in wrapper
    return func(*args, **kwargs)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 733, in rpc_sync
    fut = _invoke_rpc(to, func, RPCExecMode.SYNC, args, kwargs, timeout)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/api.py", line 643, in _invoke_rpc
    (pickled_python_udf, tensors) = _default_pickler.serialize(
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/distributed/rpc/internal.py", line 171, in serialize
    p.dump(obj)
  File "/data/users/wayi/fbsource/fbcode/buck-out/dev/gen/caffe2/test/distributed/rpc/process_group_agent#binary,link-tree/torch/jit/_script.py", line 663, in __getstate__
    raise pickle.PickleError(
_pickle.PickleError: ScriptModules cannot be deepcopied using copy.deepcopy or saved using torch.save. Mixed serialization of script and non-script modules is not supported. For purely script modules use my_script_module.save(<filename>) instead.
```

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

Commit:
Update on "[RPC Framework] Add _script_module_reducer unconditionally for RecursiveScriptModule in RPC pickler"


Previously there is no RPC pickler for `RecursiveScriptModule`. Although it is a subclass of `ScriptModule`, the reducer of `ScriptModule` is not triggered for `RecursiveScriptModule` when a script remote module is sent over RPC.

This PR checkpoints the investigation of #58274, which makes sure that a RPC pickler is invoked here. This still cannot fix `test_send_remote_module_over_the_wire_script`. Will revisit this bug once there is a feature request from users.

Differential Revision: [D28346758](https://our.internmc.facebook.com/intern/diff/D28346758/)

[ghstack-poisoned]

rohan-varma reviewed (approved):
Per offline discussion, the issue is that that RRef pickling in JIT fails because we're not in the JIT RPC context where we're serializing a message to be sent over RPC. Will continue the investigation as needed by use cases.

pritamdamania87 commented on torch/distributed/rpc/internal.py:
```diff
@@ -113,15 +113,21 @@ def serialize(self, obj):
-
-        # Add dispatch pickling for ScriptModule if needed.
+        # Add dispatch pickling for ScriptModule or its subclass.
         if isinstance(obj, torch.jit.ScriptModule):
```
Doesn't simply removing this `if` statement and always registering the reducer in the dispatch table solve this problem? Do we actually need a separate entry for RecursiveScriptModule in the dispatch table? If we do need two separate entries, I guess we can just add them unconditionally and still remove this if-statement?
